### PR TITLE
Adding release flag to cargo test

### PR DESCRIPTION
**Pull request by psdh**
#691

### CODE REVIEW

**Comment by psdh** [comment]
@alexcrichton I have made some naive changes to the cargo_test_simple, to make cargo_test_release. Please comment on how correct and complete the cargo_test_release is. After that I think I will have to write another test for verbose release testing, right?

File: tests/test_cargo_test.rs [hunk 1]
```diff
@@ -46,6 +46,44 @@ test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
...
+                assert_eq!(hello(), "hello")
+            }"#);
+
+    assert_that(p.cargo_process("build").argv("--release"), execs());
+    assert_that(&p.bin("foo"), existing_file());
+
+    assert_that(process(&p.bin("foo")).unwrap(),
+                execs().with_stdout("hello\n"));
+
+    assert_that(p.cargo("test").argv("--release"),
```
Comment (alexcrichton) [inline tests/test_cargo_test.rs]: Could you add the `-v` flag here to ensure that targets are compiled with the right flags? It's also fine to omit the above assertions around `cargo build` and wehther the binary `foo` exists.

File: tests/test_cargo_test.rs [hunk 2]
```diff
@@ -46,6 +46,41 @@ test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
...
+
+            #[test]
+            fn test_hello() {
+                assert_eq!(hello(), "hello")
+            }"#);
+
+    assert_that(process(&p.bin("foo")).unwrap(),
+                execs().with_stdout("hello\n"));
+
+    assert_that(p.cargo("test").argv("-v").argv("--release"),
```
Comment (tamird) [inline tests/test_cargo_test.rs]: this should be `.arg` rather than `.argv`

File: tests/test_cargo_test.rs [hunk 3]
```diff
@@ -46,6 +46,41 @@ test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
...
+                println!("{}", hello())
+            }
+
+            #[test]
+            fn test_hello() {
+                assert_eq!(hello(), "hello")
+            }"#);
+
+    assert_that(process(&p.bin("foo")).unwrap(),
+                execs().with_stdout("hello\n"));
```
Comment (alexcrichton) [inline tests/test_cargo_test.rs]: I think this assertion may fail as the executable hasn't been built yet?

File: tests/test_cargo_test.rs [hunk 4]
```diff
@@ -46,6 +46,41 @@ test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
...
+
+    assert_that(p.cargo("test").arg("-v").arg("--release"),
+                execs().with_stdout(format!("\
+{} foo v0.5.0 ({})
```
Comment (alexcrichton) [inline tests/test_cargo_test.rs]: You may want to be sure to run this test via `make test only=cargo_test_release` as the output has probably changed with the `-v` flag!

Comment (psdh) [inline tests/test_cargo_test.rs]: Not sure what you mean by 'run this test via `make test ...`'

File: tests/test_cargo_test.rs [hunk 5]
```diff
@@ -46,6 +46,41 @@ test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
         RUNNING)));
 });
 
+test!(cargo_test_release {
+    let p = project("foo")
+        .file("Cargo.toml", basic_bin_manifest("foo").as_slice())
```
Comment (tamird) [inline tests/test_cargo_test.rs]: `as_slice()` is deprecated (and failing your travis build). you'll want to rebase and change this to `&basic_bin_manifest("foo")`

Comment (tamird) [inline tests/test_cargo_test.rs]: consider squashing your commits while you're rebasing

File: tests/test_cargo_test.rs (hunk 4 above)
Comment (alexcrichton) [inline tests/test_cargo_test.rs]: Oh that's just the command to use to run just this tests.

**Comment by alexcrichton** [comment]
@psdh there's a [failure on travis](https://travis-ci.org/rust-lang/cargo/jobs/57070835#L943) with the test that was added that should be addressed.

**Comment by psdh** [comment]
I have been trying to figure out the failure. It somehow manages to not execute the `cargo test -v --release` command at all (right?). I am not able to see how it fails. Do you have any suggestions?

**Comment by alexcrichton** [comment]
You'll want to change `p.cargo(...)` to `p.cargo_process(...)`

**Comment by psdh** [comment]
@alexcrichton do we need more tests?

**Comment by alexcrichton** [comment]
Oh well I was hoping that this would test for flags being passed like `-C opt-level=3` as well as optimizing dependencies (and the tests themselves). Could you expand the test case to have a library, an integration test, and a dependency, all of which should be compiled with optimizations?

**Comment by psdh** [comment]
Sorry for the delay in getting this through, Could you point me towards some example of a similar test case.

**Comment by alexcrichton** [comment]
You can see examples of matching `-C opt-level` in [`test_cargo_bench.rs`](https://github.com/rust-lang/cargo/blob/master/tests/test_cargo_bench.rs#L661-L757) but the test case probably doesn't need to be as large.

**Pull request merged**